theano/sandbox/cuda/tests/test_cusolver.py

```diff
@@ -0,0 +1,28 @@
+import numpy
+import theano
+from numpy.testing import assert_allclose
+from theano import tensor
+from theano.sandbox.cuda.cusolver import geqrf
+from scipy.linalg.lapack import dgeqrf
+
+
+def test_geqrf():
+    rng = numpy.random.RandomState(1)
+
+    def assert_geqrf_allclose(m, n):
+        A = rng.rand(m, n).astype(theano.config.floatX)
+        A_tensor = tensor.matrix('A')
+        dgeqrf_ = theano.function([A_tensor], geqrf(A_tensor))
+        qr_, tau_, work_, info_ = dgeqrf_(A)
+        qr, tau, work, info = dgeqrf(A)
+        assert_allclose(qr, qr_, 1e-3)
+        assert_allclose(tau, tau_, 1e-3)
+        print(m, n, 'correct')
```

Comment:
That's to make sure we mirror `scipy.linalg.lapack.sgeqrf`. I think it
would be good if their interfaces are the same, and SciPy returns it,
although I don't think it's actually very useful.

On Wed, Jun 10, 2015 at 12:59 PM, Frédéric Bastien <notifications@github.com
> wrote:

> In theano/sandbox/cuda/tests/test_cusolver.py
> <https://github.com/Theano/Theano/pull/3013#discussion_r32139950>:
>
> > +from theano.sandbox.cuda.cusolver import geqrf
> > +from scipy.linalg.lapack import dgeqrf
> > +
> > +
> > +def test_geqrf():
> > +    rng = numpy.random.RandomState(1)
> > +
> > +    def assert_geqrf_allclose(m, n):
> > +        A = rng.rand(m, n).astype(theano.config.floatX)
> > +        A_tensor = tensor.matrix('A')
> > +        dgeqrf_ = theano.function([A_tensor], geqrf(A_tensor))
> > +        qr_, tau_, work_, info_ = dgeqrf_(A)
> > +        qr, tau, work, info = dgeqrf(A)
> > +        assert_allclose(qr, qr_, 1e-3)
> > +        assert_allclose(tau, tau_, 1e-3)
> > +        print(m, n, 'correct')
>
> Why return the working memory? Do it contain useful information? Can it be
> reused in other computation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3013/files#r32139950>.
>